python/src/iceberg/types.py

```diff
@@ -93,21 +121,37 @@ def precision(self) -> int:
     def scale(self) -> int:
         return self._scale
 
-    def __eq__(self, other):
-        if type(self) is type(other):
-            return self.precision == other.precision and self.scale == other.scale
-        return False
 
+class NestedField(IcebergType):
+    """This represents a field of a struct, a map key, a map value, or a list element. This is where field IDs, names, docs, and nullability are tracked."""
+
+    _instances: Dict[Tuple[bool, int, str, IcebergType, Optional[str]], "NestedField"] = {}
+
+    def __new__(
+        cls,
+        is_optional: bool,
```

Comment:
The optional boolean is at the end for map and list types. What about adding this after `doc` and defaulting it to `True`?